### PR TITLE
firewall: Allow editing custom services

### LINKED ISSUE #12287: No ability to edit firewall services
Cockpit version: 197
OS: Fedora Server 30
Page: Network > Firewall

In the usability test we ran now in July with 9 participants, 8 out of 9 participants wanted to be able to edit a Firewall service without having have to delete and creating a new one.

Frequency: 88,89%
Severity: 5/5

### PR DESCRIPTION
**Pull request by marusak**
Fixes #12287

Screenshot from 2022-06-07 13-07-56

Screenshot from 2022-06-07 13-08-11

### CODE REVIEW

File: test/verify/check-networkmanager-firewall [hunk 1]
```diff
@@ -409,6 +409,19 @@ class TestFirewall(NetworkCase):
         # should have been removed in the reload
         b.wait_not_present("tr[data-row-id='http']")
 
+        b.click("#add-services-dialog button.btn-cancel")
```
Comment (martinpitt) [inline test/verify/check-networkmanager-firewall:412]: This belongs after the `wait_in_text` in line 407 IMHO.

File: test/verify/check-networkmanager-firewall [hunk 2]
```diff
@@ -409,6 +409,19 @@ class TestFirewall(NetworkCase):
...
+        b.click("tr[data-row-id='custom--80-82-echo-83-123-snmp'] button[aria-label=Actions]")
+        b.click("a:contains(Edit)")
+        b.wait_val("#tcp-ports", "80-82, 7")
+        b.wait_val("#udp-ports", "83-123, 161")
+        b.wait_val("#service-name", "custom--80-82-echo-83-123-snmp")
+        b.wait_val("#service-description", "custom snmp firewall rule")
+        b.set_input_text("#tcp-ports", "80-82, 8")
+        b.set_input_text("#service-description", "edited snmp firewall rule")
+        b.wait_val("#service-name", "custom--80-82-echo-83-123-snmp")
+        save("custom--80-82-echo-83-123-snmp", "80-82, 8", "83-123, 161", "edited snmp firewall rule", 'Edit service')
```
Comment (martinpitt) [inline test/verify/check-networkmanager-firewall:423]: Does that validate the actual settings in firewalld anywhere? That tcp/7 is really closed, tcp/8 got opened, and UDP survived?

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thank you!

File: pkg/networkmanager/firewall-client.js [hunk 3]
```diff
@@ -396,6 +396,28 @@ firewall.createService = (service, ports, zones, desc = "") => {
...
+ *
+ * Returns a promise that resolves when the service is edited.
+ */
+firewall.editService = (service, ports, desc = "") => {
+    return firewalld_dbus.call('/org/fedoraproject/FirewallD1/config',
+                               'org.fedoraproject.FirewallD1.config',
+                               'getServiceByName', [service])
+            .then(path => firewalld_dbus.call(path[0],
+                                              'org.fedoraproject.FirewallD1.config.service',
+                                              'update2', [{ description: { t: 's', v: desc }, ports: { t: 'a(ss)', v: ports } }])
```
[the remark this answers is not in the document]
Comment (marusak) [reply pkg/networkmanager/firewall-client.js:410]: Yes it does. You give it a new value for ports and it changes it to them. Whether you add, remove or change completely, it just applies the new ports.

**Comment by marusak** [comment]
`RuntimeError: TypeError: Cannot read properties of null (reading '8')` oopses sometimes on [rhel](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17436-20220608-072422-0f7cc838-rhel-8-7/log.html#305-2). I will need to debug this.

**Review by garrett — DISMISSED** [review]
The title is missing the word "custom". So the title would be this instead:

Edit custom service in Fedora Workstation Zone

---

Additionally: That alert should be a helpertext. (Alerts have to be at the top of modals, not at the bottom, but alerts should be reactive not proactive. Proactive text is anticipatory and should be helper text at the bottom. In other words: The ahead-of-time warning text has the right idea in the right place, just the wrong widget type.)

This warning text change could either be a separate commit or a follow-up PR.

**Comment by marusak** [comment]
> The title is missing the word "custom". So the title would be this instead:
> Edit custom service in Fedora Workstation Zone

Done

> `RuntimeError: TypeError: Cannot read properties of null (reading '8')` oopses sometimes on [rhel](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17436-20220608-072422-0f7cc838-rhel-8-7/log.html#305-2).

Figured it out and fixed.

> Additionally: That alert should be a helpertext

Right. I haven;t touched that in this PR. Not super sure how it should look like so I don't want to block this PR on it. Lets adjust it outside of this PR.

**Review by martinpitt — APPROVED** [review]
Yay, dakujem! Can you please update the screenshot and add a release note?

**Pull request merged**